utils/src/resource_location.cpp

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2022 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+
+#include "utils/resource_location.h"
+
+#include <filesystem>
+#include <regex>
+
+#include "network/uri/uri.hpp"
+
+#include "utils/path.h"
+#include "utils/path_conversions.h"
+#include "utils/platform.h"
+
+namespace hlasm_plugin::utils::resource {
+
+struct dissected_uri
+{
+    std::optional<std::string> scheme = std::nullopt;
+    std::optional<std::string> auth = std::nullopt;
+    std::optional<std::string> path = std::nullopt;
+    std::optional<std::string> query = std::nullopt;
+    std::optional<std::string> fragment = std::nullopt;
+};
+
+std::string get_file_path(const network::uri& u)
+{
+    if (!u.has_path())
+        return "";
+
+    network::string_view path = u.path();
+
+    if (utils::platform::is_windows())
+    {
+        // we get path always beginning with / on windows, e.g. /c:/Users/path
+        path.remove_prefix(1);
+    }
+
+    std::string auth_path = path.to_string();
+    if (utils::platform::is_windows())
+    {
+        auth_path[0] = (char)tolower((unsigned char)auth_path[0]);
+    }
+
+    return utils::path::lexically_normal(network::detail::decode(auth_path)).string();
+}
+
+dissected_uri dissect_uri(const std::string& uri)
+{
+    dissected_uri ret;
+
+    try
+    {
+        network::uri u(uri);
+
+        if (!u.scheme().compare("file") && (!u.has_authority() || u.authority().to_string() == ""))
+        {
+            ret.path = get_file_path(u);
+            return ret;
+        }
+
+        if (u.has_scheme())
+            ret.scheme = u.scheme().to_string();
+        if (u.has_authority())
+            ret.auth = u.authority().to_string();
+        if (u.has_path())
+            ret.path = u.path().to_string();
+        if (u.has_query())
+            ret.query = u.query().to_string();
+        if (u.has_fragment())
+            ret.fragment = u.fragment().to_string();
+
+        return ret;
+    }
+    catch (const std::exception&)
+    {
+        return ret;
+    }
+}
+
+resource_location::resource_location(std::string uri)
+    : m_uri(std::move(uri))
+{}
+
+resource_location::resource_location(std::string_view uri)
+    : resource_location(std::string(uri))
+{}
+
+resource_location::resource_location(const char* uri)
+    : resource_location(std::string(uri))
+{}
+
+const std::string& resource_location::get_uri() const { return m_uri; }
+
+std::string resource_location::get_path() const { return m_uri.size() != 0 ? utils::path::uri_to_path(m_uri) : m_uri; }
+
+std::string resource_location::to_presentable() const
+{
+    dissected_uri dis_uri = dissect_uri(m_uri);
```

Comment:
Keep this code as well for debugging, but I guess add an argument to the function (defaulting to simple human readable form).

As a user, I don't want to see `file://c%3A/folder%20with%20spaces/file%20name` in my messages, and while seeing `file://c:/folder with spaces/file name` might be ok, I'd certainly expect just "c:\folder with spaces\file name"